Bluetooth: host: Improve HCI command send error handling

joerchan opened the pull request:
Handle HCI command send sync timeout by reporting the error back to the
application and logging a warning.
Log command status failure with information in order to make it more
visible as the HCI status code is more interesting than the -EIO error
code returned by the function.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Use bluetooth assert on HCI command send error

Use bluetooth assert on HCI command send error since this assertion is
always enabled and we should not continue after this has failed.
Log command status failure with information in order to make it more
visible as the HCI status code is more interesting than the -EIO error
code returned by the function.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -364,31 +365,41 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
-	BT_DBG("opcode 0x%04x status 0x%02x", opcode, cmd(buf)->status);
+	status = cmd(buf)->status;
+	if (status) {
+		BT_INFO("opcode 0x%04x status 0x%02x", opcode, status);
```
I think BT_WARN() would be more appropriate

joerchan replied:
Updated to BT_WARN

jhedberg reviewed (changes requested):
Looks generally good, just a small tweak to the logging

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -364,31 +365,41 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
 	err = k_sem_take(&sync_sem, HCI_CMD_TIMEOUT);
-	__ASSERT(err == 0, "k_sem_take failed with err %d", err);
+	if (err) {
+		BT_WARN("Failed to send command to driver (%d)", err);
```
This should be `BT_ERR()` (sorry if I missed it previously). Basically, `BT_ERR()` should be used when there's a definitive error condition coming from outside of our subsystem or component (the HCI driver or controller in this case). `BT_WARN()` OTOH is for cases where there's not necessarily an error, but might be, and it's important to make the developer/user aware of this. `BT_INFO()` should be reserved for rare circumstances where we have some generally informational content to show the developer/user - e.g. that's why it's used to log the basic controller information upon HCI init.

jhedberg replied:
Actually, putting the logging details aside: why did you remove the assert and try to handle this at runtime? Isn't the HCI command flow control broken if we fail to wait for the command complete/status? IIRC that's why this was an assert.

jhedberg replied:
IIRC originally this was with K_FOREVER, so neither assert nor BT_ERR() was useful. We later introduced the HCI_CMD_TIMEOUT so we could log this condition, but it doesn't change the fact that something is critically broken if this happens and we probably shouldn't attempt to continue normal operation of the stack.

jhedberg replied:
Also, the log message is incorrect: we *did* succeed sending the command to the driver, but then subsequently timed out waiting for a command complete/status.

joerchan replied:
At first I wanted to add an BT_ASSERT instead of __ASSERT, since as you say something is broken, and the __ASSERT can be disabled. However I thought it was too harsh to crash the host in a split build where you are communicating with an HCI controller that might have crashed.

But you are right, something is critically broken. Maybe we could unset the BT_DEV_READY flag, send an HCI reset command, but unsure if that handles everything.

Maybe a BT_ASSERT is the best option after all.

jhedberg replied:
yeah, I'd go with that as well

joerchan replied:
Added BT_ASSERT_MSG to keep the message.

joerchan committed:
Bluetooth: log: Add BT_ASSERT_MSG that matches __ASSERT

Add BT_ASSERT_MSG assert mechanism that can print assertion messages
when verbose bluetooth asserts are enabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
